### DISCUSSION TITLE
Symfony Messenger with RabbitMQ JsonSerialize problem with requeue

### DISCUSSION THREAD

**Discussion by darius1706** [symfony/symfony#49399] [Q&A]
Hello,
I have two separate systems that communicate with each other via RabbitMQ.
The first one sends a array in JsonFormat to the queue, the second one should pick up the message and process it.
I would like to use Symfony Messenger for processing messages, which, however, needs deserialization to receive a message in Json format.
This class for JsonSerialization, which is taken from the Symfony Messenger documentation, works fine until an exception is thrown and the message should be returned back to the RabbitMQ queue.
In that case, the message `[Exception] Serialization of 'Closure' is not allowed` is thrown and the failure occurs on the line

`'stamps' => serialize($allStamps),`

in function encode.

```
class RabbitMessageMessengerJsonSerializer implements SerializerInterface
{
    public function decode(array $encodedEnvelope): Envelope
    {
        $body = $encodedEnvelope['body'];
        $headers = $encodedEnvelope['headers'];
        $data = json_decode($body, true);
        $message = new RabbitMessageMessenger($data);

        $envelope = new Envelope($message);

        $stamps = [];
        if (isset($headers['stamps'])) {
            $stamps = unserialize($headers['stamps']);
        }

        return $envelope->with(... $stamps);
    }

    public function encode(Envelope $envelope): array
    {
        $message = $envelope->getMessage();
        if ($message instanceof RabbitMessageMessenger) {
            $data = $message->getContent();
        } else {
            throw new \Exception('Unsupported message class');
        }

        $allStamps = [];
        foreach ($envelope->all() as $stamps) {
            $allStamps = array_merge($allStamps, $stamps);
        }

        return [
            'body' => json_encode($data),
            'headers' => [
                'stamps' => serialize($allStamps),
            ],
        ];
    }
}

```
Thank you for your help

**Comment by ghost** [comment]
Please reformat the code sample because it's broken. Please uses PHP code block. Please point explicitly line 46, because nobody will count the lines up to 46. I would also recommend to improve the description because it hard to read and understand. Please use Google translator if you're not confident about your English.
Remember, the better and more friendly to read your issue description is, the more people will be willing to read and answer this.

**Reply by ghost** [reply]
Please update the Discussion description and don't describe your issue again in the reply to my comment.

**Reply by ghost** [reply]
If your message is in JSON format why do you use `serialize()` and `unserialize()` methods? Or is it JSON but the `$msg['headers']['stamps']` field is serialized by the producer using PHP's `serialize()` function? Are both application using Symfony Messenger? What is the structure of the message that is being exchanged between application?
Plase provide more details.

**Reply by darius1706** [reply]
Good questions, but this is clearly how Symfony Messenger is designed, at least in the documentation and sample example. It says there that in case of failure, the message must be made again, which is then returned to the RabbitMQ queue by means of Messenger.
This implementation is taken from the official link https://symfony.com/doc/current/messenger.html , which further links to: https://symfonycasts.com/screencast/messenger/transport-serializer
Both apps don't use Symfony Messenger, but I don't see why that should be a problem, that's why we're using RabbitMQ after all.
Yes, I'm considering moving to RabbitMqBundle here as well, but I use Messenger here for other things too, it would be unnecessary to have another tool that does the same thing.
The message structure is a simple array encoded in Json. Here is example:
`{"messageType":"update","domain":"xxx.cz","domainId":1}`
Here, for one, there is a problem with the stamps, what I wrote down is a lot of control information, for example, about the failure of the program, about how the requey should be performed, etc.
I tried to remove some stamps, unfortunately this error did not occur, but Messenger started returning messages to the queue as a copy of itself at high speed.

**Reply by ghost** [reply]
Please read the conversations under the link to symfony casts. Maybe someone else had this problem and you will find a hint.
You can also use xDebug to see how the stamps from`$allStamps` look like. See if any holds a closure somewhere.

**Reply by darius1706** [reply]
Thank you, I will try and will write result.

**Reply by darius1706** [reply]
Unfortunately, I was unable to solve the problem with any advice or instructions I found. When I follow the instructions strictly and it's just Messenger and one queue, it never seems to throw the error. However, as soon as the definition of queues with x-dead-letter etc. enters into it, the aforementioned problem with serialization occurs.
I can't dwell on it any longer at this point, I'm leaving Messenger and going back to the php-amqplib/rabbitmq-bundle library, I'll hope that sometime in the future the library authors fix the problem and I don't have to use two things for the same thing.
